### PR TITLE
[quant][core][feature] Implemented index_put for QuantizedCPU tensors

### PR DESCRIPTION
**Pull request by dzdang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78384
* #78368

Summary:
Previously, there was no support for index_put for quantized tensors.
This PR introduces the feature for quantized CPU tensors, but is limited
to per tensor quantized tensors and accumulate=False. A
corresponding test case was added to test the impl.

Most of the implementation is copied over from the fp kernel and adapted
for quantized tensors.

TODO: implement index_put for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put
```

Differential Revision: [D36716641](https://our.internmc.facebook.com/intern/diff/D36716641)

### CODE REVIEW

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36716641).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36716641).

File: aten/src/ATen/native/quantized/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -60,5 +77,43 @@ Tensor & masked_fill__quantized_cpu(Tensor& self, const Tensor & mask, const Ten
   return self;
 }
 
+Tensor& _index_put_impl_quantized_cpu_(Tensor & self, const torch::List<c10::optional<Tensor>>& indices, const Tensor & value, const bool accumulate, const bool unsafe) {
+  TORCH_CHECK_INDEX(indices.size() <= (size_t)self.dim(), "too many indices for tensor of dimension ", self.dim(), " (got ", indices.size(), ")");
+  TORCH_CHECK(!value.is_quantized(), "Value argument for quantized input_put should not be quantized");
+  TORCH_CHECK(self.qscheme() == c10::kPerTensorAffine, "index_put for quantized tensors is currently only supported for per tensor quantized tensors");
+  TORCH_CHECK(!accumulate, "index_put for quantized tensors is currently only supported for accumulate=False");
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/TensorAdvancedIndexing.cpp:84]: it's unexpected to throw this exception but also have the accumulate branches in the code, can we pick just one of these but not both?

Comment (dzdang) [reply aten/src/ATen/native/quantized/TensorAdvancedIndexing.cpp:84]: will do

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36716641).

**Review by vkuzo — APPROVED** [review]

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36716641).

**Pull request closed without merging**